File: dsp22.c
#include <graph.h>
#include <conio.h>
#include <stdio.h>
#include <stdlib.h>
#include <math.h>

#define DIM     320

float fX[DIM + 1], fY[DIM + 1], fP[DIM + 1], fQ[DIM + 1],
      fR[DIM + 1], fS[DIM + 1], fT[DIM + 1], fU[DIM + 1],
      fV[DIM + 1], fW[DIM + 1];

void main (void)
    {
    /*
     *** PROGRAM NO. 22    FREQUENCY-SAMPLING FILTER ***********
     */
    int    iCount, iCount2, iM;
    float  fFREQ, fHI, fH, fHR;

    _clearscreen  (_GCLEARSCREEN);
    printf ("PROGRAM 22: Frequency sampling filter\n");

    fX[120] = 1.0;

    /*
     ************** LOAD IMPULSE INTO INPUT ARRAY **************
     */
    _setvideomode( _MRES16COLOR );

    for (iCount = 120; iCount <= 280; iCount++)
        {
        fW[iCount] = fX[iCount] - 0.886867 * fX[iCount-120];
        fP[iCount] = 0.5171211 * fP[iCount-1] - 0.998001 *
                     fP[iCount-2] + 0.5 * fW[iCount];
        fQ[iCount] = 0.415408 * fQ[iCount-1] - 0.998001 *
                     fQ[iCount-2] - fW[iCount];
        fR[iCount] = 0.312556 * fR[iCount-1] - 0.998001 *
                     fR[iCount-2] + fW[iCount];
        fS[iCount] = 0.208848 * fS[iCount-1] - 0.998001 *
                     fS[iCount-2] - fW[iCount];
        fT[iCount] = 0.104567 * fT[iCount-1] - 0.998001 *
                     fT[iCount-2] + fW[iCount];
        fU[iCount] = -0.998001 * fU[iCount-2] - 0.666667 *
                     fW[iCount];
        fV[iCount] = -0.104567 * fV[iCount-1] - 0.998001 *
                     fV[iCount-2] + 0.333333 * fW[iCount];
        fY[iCount] = fP[iCount] + fQ[iCount] + fR[iCount] +
                     fS[iCount] + fT[iCount] + fU[iCount] +
                     fV[iCount];

        /*
         *************** PLOT IMPULSE RESPONSE VALUE ***********
         */
        iM = (iCount - 120) * 2;
        _moveto (iM, 70);
        _lineto (iM, (int)(70 - fY[iCount] * 7.0 + 0.5));
        }

    for (iCount = 1; iCount <= 320; iCount++)
        {
        fFREQ = (float) (iCount - 1) * 3.141593 / 320.0;
        fHR = 0.0;
        fHI = 0.0;
        fH = 0.0;

        for (iCount2 = 1; iCount2 <= 120; iCount2++)
            {
            fHR += fY[119 + iCount2] * cos ((float) iCount2 *
                fFREQ);
            fHI += fY[119 + iCount2] * sin ((float) iCount2 *
                fFREQ);
            }

        fH = pow ((fHR * fHR + fHI * fHI), 0.5);

        /*
         ************* PLOT FREQUENCY RESPONSE VALUE ***********
         */
        _moveto (iCount, 170);
        _lineto (iCount, (int)(170 - 0.75 * fH + 0.5));
        }

    getch();
    _setvideomode( _DEFAULTMODE );
    
    /*
     ***********************************************************
     */
    }

File: dsp22.pas
(********  PROGRAM NO.22    FREQUENCY-SAMPLING FILTER  *********)
program DSP22;
uses graph3;
var X: array[1..320] of real;
var Y: array[1..320] of real;
var W: array[1..320] of real;
var P: array[1..320] of real;
var Q: array[1..320] of real;
var R: array[1..320] of real;
var S: array[1..320] of real;
var T: array[1..320] of real;
var U: array[1..320] of real;
var V: array[1..320] of real;
var N,M,K: integer;
var FREQ,HR,HI,H: real;
begin
  for N:=1 to 320 do
    begin
      X[N]:=0;Y[N]:=0;W[N]:=0;P[N]:=0;Q[N]:=0;
      R[N]:=0;S[N]:=0;T[N]:=0;U[N]:=0;V[N]:=0;
    end;
  graphmode;
(**************  LOAD IMPULSE INTO INPUT ARRAY  ****************)
  X[120]:=1;
    for N:=120 to 280 do
      begin
        W[N]:=X[N]-0.886867*X[N-120];
        P[N]:=0.5171211*P[N-1]-0.998001*P[N-2]+0.5*W[N];
        Q[N]:=0.415408*Q[N-1]-0.998001*Q[N-2]-W[N];
        R[N]:=0.312556*R[N-1]-0.998001*R[N-2]+W[N];
        S[N]:=0.208848*S[N-1]-0.998001*S[N-2]-W[N];
        T[N]:=0.104567*T[N-1]-0.998001*T[N-2]+W[N];
        U[N]:=-0.998001*U[N-2]-0.666667*W[N];
        V[N]:=-0.104567*V[N-1]-0.998001*V[N-2]+0.333333*W[N];
        Y[N]:=P[N]+Q[N]+R[N]+S[N]+T[N]+U[N]+V[N];
(***************  PLOT IMPULSE RESPONSE VALUE  *****************)
        M:=(N-120)*2;
        draw(M,70,M,round(70-Y[N]*7),1);
      end;
  for N:=1 to 320 do
    begin
      FREQ:=(N-1)*3.141593/320;HR:=0;HI:=0;H:=0;
        for K:=1 to 120 do
          begin
            HR:=HR+Y[119+K]*cos(K*FREQ);
            HI:=HI+Y[119+K]*sin(K*FREQ);
          end;
        H:=sqrt(HR*HR+HI*HI);
(**************  PLOT FREQUENCY RESPONSE VALUE  ****************)
      draw(N,170,N,round(170-0.75*H),1);
    end;
end.
(***************************************************************)